### PR TITLE
ENH: Replaces ipyparallel with multithreading

### LINKED ISSUE #92: Possibly look into the Python builtin `multiprocessing` for parallelization?
Ran into a dependency issue on a plane recently, where I didn't have `ipyparallel` installed (development environment), so couldn't utilize parallelization. Barring any hard dependencies with `ipyparallel` it seems it'd make sense to swap for native python parallelization through the [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html) package.

I could kick up a benchmark to see if there are any gains or losses between the two.

### LINKED ISSUE #98: Hangup signal lost
When running sourcetracker2 I get this warning message: 

/home/gabif/miniconda3/envs/st2/lib/python3.5/site-packages/ipyparallel/client/client.py:458: RuntimeWarning:
            Controller appears to be listening on localhost, but not on this machine.
            If this is true, you should specify Client(...,sshserver='you@mac-nugget')
            or instruct your controller to listen on an external IP.
  RuntimeWarning)

Everything runs fine, the job completes and the files get written, but this connection doesn't get hung up (and I need to ctrl-C it once checking the output is there).

### LINKED ISSUE #106: Sourcetracker2 runs for 1m30s then all processes switch to 'Sleep' and hangs
Hello,

**Issue**
I just tried running Sourcetracker2, as installed via instructions on the README (anaconda then pip install).

When I run the program with the `--jobs` parameter, and monitor via htop, the program runs for 1m30s and then stops, essentially hanging. A bunch of sourcetracker processes remain sitting with the 'sleep' status, and sourcetracker2 runs indefinitely (i.e. never completes - I left it for 24 hours in one case).

**Description**
To describe what I did:

1) Get the github repo to use the tiny-test data `git clone https://github.com/biota/sourcetracker2.git`
1) Installed Anaconda3.-5.1.0.Linux-x86_64.sh
2) Made the st2 env `conda create -n st2 -c biocore python=3.5 numpy scipy scikit-bio biom-format h5py hdf5 seaborn`
3) Activated the env `source activate st2`
4) Installed sourcetracker `pip install sourcetracker `
5) Changed to the tiny-test directory
6) Run command with no --jobs (this completes successfully) `sourcetracker2 gibbs -i otu_table.biom -m map.txt -o example1/`
7) Run command with `--jobs 5`, which hangs `sourcetracker2 gibbs -i otu_table.biom -m map.txt -o example1/ --jobs 5`

I attach two screenshots htop during the `--jobs 5` run, both pre- and post 1 minute 30 seconds. Note the pre-1:30 has many more st2 related processes but I've cropped for space reasons.

01-st2_multijobs_running_pre1m30

02-st2_multijobs_hanging_post1m30

I have noted that individual files are generated for each sample in the results directory (s0.txt, s1.txt, s2.txt etc.). I also tried with and without modifying the `--cluster-delay-start` parameter with no differences.

As a sidenote, it would be nice of some stdout and stderr messages could be added, as it would help me investigate further and be more precise in my error reporting.

**Environment Information** 
The sourcetracker2 environment information:

_Anaconda_
 conda environments:

base                  *  /projects1/users/fellows/bin.backup/anaconda3
st2                      /projects1/users/fellows/bin.backup/anaconda3/envs/st2

_Sourcetracker2_
[st2_env_info.txt](https://github.com/biota/sourcetracker2/files/1963631/st2_env_info.txt)

### PR DESCRIPTION
**Pull request by johnchase**
fixes #106, fixes #98, fixes #92
This pull request replaces ipyparallel with the python multithreading
library. This should prevent some of the issues that we have seen
running sourcetracker in parallel previously

### CODE REVIEW

**Comment by wdwvt1** [comment]
@johnchase - I'll take a look at this over the next 2 days!

**Comment by johnchase** [comment]
Thanks @jakereps I updated the travis settings so that it will show build status on the prs (although, now it's kicking off two pipelines). I pulled out the `Pool` code so it's only in one place, and added a test with two processors. I did not update tests to pass with python > 3.5 as I think that may be a bigger discussion

File: sourcetracker/_sourcetracker.py [hunk 1]
```diff
@@ -728,24 +734,19 @@ def gibbs(sources, sinks=None, alpha1=.001, alpha2=.1, beta=10, restarts=10,
...
-    >>> time.sleep(25)
-    >>> c = Client()
+    # Run the function on multiple processors
     >>> mpm, mps, fas = gibbs(source_df, sink_df, alpha1, alpha2, beta,
                               restarts, draws_per_restart, burnin, delay,
-                              cluster=c, create_feature_tables=True)
+                              jobs=4, create_feature_tables=True)
 
     # LOO prediction.
     >>> mpm, mps, fas = gibbs(source_df, sinks=None, alpha1, alpha2, beta,
```
Comment (wdwvt1) [inline sourcetracker/_sourcetracker.py]: This examples doesn't work (your PR didn't break it I don't think). Needs to read 
```python
    # LOO prediction.
    >>> mpm, mps, fas = gibbs(source_df, sinks=None, alpha1=alpha1,
                              alpha2=alpha2, beta=beta, restarts=restarts,
                              draws_per_restart=draws_per_restart,
                              burnin=burnin, delay=delay, jobs=5,
                              create_feature_tables=True)
```

**Comment by wdwvt1** [comment]
Do you want to squash and merge or should we discuss the `pyhton 3.7` changes that would be necessary?

**Comment by johnchase** [comment]
Thanks @wdwvt1! 

> should we discuss the pyhton 3.7 changes that would be necessary?

I think this should be a separate issue, although likely a worthwhile

**Comment by jkliu** [comment]
Branch works fine for me, except when I define jobs too high. 
README.md still mentions ipyparallel.

**Comment by johnchase** [comment]
> Branch works fine for me, except when I define jobs too high.

@jkliu Can you be specific about what doesn't work, and what too many jobs are?

**Comment by jkliu** [comment]
When I define 581 jobs, SourceTracker doesn't finish in a couple hours.  I would also get some sort of "file save" error from jupyter hub (didn't take a screenshot).

**Comment by wdwvt1** [comment]
@johnchase thanks for all the work, I merged this branch and added the issue #119 to deal with the too many jobs issue raised by @jkliu.

**Pull request merged**